### DISCUSSION TITLE
How you manage automatic project setup with SSO/RBAC at scale?

### DISCUSSION THREAD

**Discussion by silveiralexf** [argoproj/argo-workflows#11779] [Q&A] [answered]
Hi there folks, I have a question regarding what type of strategy you are using for access control for hundreds of users and/or thousands of projects (each with different permission levels) with SSO (LDAP).

Right now I'm working on an alternative for a couple of very large Jenkins instances, where we have thousands of pipelines running every day, every time developers push changes on their repos, and a large number of infrastructure pipelines for provisioning resources, cronjobs, etc.

My current use case is, around 800+ developers and something around 6k Git repositories. Each repository gets a namespace created so we can have a cleaner display on the UI, especially to "match" the multi-branch pipeline view provided by Jenkins, so users can filter their workflows per namespace/project and Pull Request(s), using different labels for different types of workflows.

Every developer has read access for all company projects (to which we have cluster roles and cluster role bindings created as part of our deployment).

However, as project maintainers must have some extra write permissions to be able to create/start/stop workflows on the namespace of their projects, and most users are maintainers of multiple projects, the only solution I could find to combine roles was having a service account to each user (on workflows namespace), mapped to their SSO user using RBAC/precedence and role bindings per namespace.

The main reason for having the requirement to map users to projects is access to credentials. We have Vault in our setup, and we want to control access to credentials through Vault roles -- therefore only authorized service accounts can reach the secrets of a given namespace (that matches the Vault credential path/role). 

The source of truth for my users is a manifests mono repo, where each file represents a project/repository with a list of maintainers. To avoid having to manually manage requests for create/update access to projects, I've created a small controller that consumes the list of maintainers and creates/update/delete the required namespaces/roles/role bindings and vault roles for the users whenever a change is detected on our the manifests mono repo.

This works well, and I have managed to automate the whole thing quite well after a couple of weeks of banging my head to put this stuff together, but I was wondering if other folks have found similar problems at scale, and how you are managing more "granular" access requirements?

Appreciate a lot any input/feedback!

Thanks in advance!

**Comment by agilgur5** [comment, marked as answer]
I think that you've implemented a variation of the best solution that is currently available.
There is a feature request for more granular RBAC with Casbin (similar to Argo CD) in #6490.

I've implemented something similar for Data Scientist permissions, where each model has its own namespace and permissions limited to the maintainers of that model.
The only real difference was that there was a separate database with these permissions that the RBAC was synced to, instead of a manifests repo.

The controller implementation is one way of implementing synchronization; another could to be to run a Workflow whenever there are changes. A controller is probably a bit more robust, though is a separate piece of software to maintain (trade-offs).

**Reply by silveiralexf** [reply]
thanks, @agilgur5 for the feedback!
